### PR TITLE
Add 'file://' prefix to sourceURL when loading a script from a file.

### PR DESCRIPTION
**Pull request by skevy**
When loading a bundle from a file (usually in the case of the release version of the application), we should prepend "file://" to the file path when passing a "sourceURL" to the JSC bridge.

This 1) unifies iOS and Android when it comes to style in which they show the file name of the script in a traceback, and 2) making this URL and _actual_ URL allows it to be parsed by something like TraceKit (https://github.com/csnover/TraceKit), which is used by Sentry, for instance, to be able to easily log tracebacks.

### CODE REVIEW

File: ReactAndroid/src/main/java/com/facebook/react/bridge/JSBundleLoader.java [hunk 1]
```diff
@@ -31,7 +31,7 @@ public void loadScript(ReactBridge bridge) {
         if (fileName.startsWith("assets://")) {
           bridge.loadScriptFromAssets(context.getAssets(), fileName.replaceFirst("assets://", ""));
```
Comment (ide) [inline ReactAndroid/src/main/java/com/facebook/react/bridge/JSBundleLoader.java]: Probably want to do the same here for consistency

Comment (skevy) [inline ReactAndroid/src/main/java/com/facebook/react/bridge/JSBundleLoader.java]: Ah. You're right.

**Comment by mkonicek** [comment]
Makes sense to me. How did you test this? Did you test both branches?

**Comment by skevy** [comment]
I did not test both branches...as I'm actually not sure how to make the "assets://" stuff work haha (but I'm happy to test if you have any ideas).

Regarding testing the other code path...really, I just ran some apps, and threw Errors. This change really only makes a change to the url that's reported in tracebacks...you can actually see a comment in this same file alluding to that fact a few lines down:

"Providing correct {@param sourceURL} of downloaded bundle is required for JS stacktraces to work correctly and allows for source maps to correctly symbolize those."

I'm just doing the same thing for the bundle that's loaded from a file in release mode.

**Comment by mkonicek** [comment]
Ah OK, you've tested only the `loadScriptFromFile` branch, correct? I don't have much context on this part of the codebase but I'd say please change only the code you were able to test.

I thought `loadScriptFromAssets` was called in release mode when the JS bundle is packaged inside the APK ([docs](https://facebook.github.io/react-native/docs/signed-apk-android.html)) and `loadScriptFromFile` was called when we're loading the JS bundle from a custom file, such as when using services like Microsoft Code Push. But might be mistaken.

**Comment by skevy** [comment]
@mkonicek I think you're probably right.

I will test this. I don't expect it to be an issue, but will test to be sure and report back.

**Comment by geof90** [comment]
I don't think the change to the "assets://" code path will work. The bridge API requires that second param to be the name of the asset, not a URL.

File: ReactAndroid/src/main/java/com/facebook/react/bridge/JSBundleLoader.java (hunk 1 above)
Comment (skevy) [inline ReactAndroid/src/main/java/com/facebook/react/bridge/JSBundleLoader.java]: Nope...turns out that function takes an "assetName" param. So we're gonna leave it alone here, and change it where the script is executed in `OnLoad.cpp`

**Comment by skevy** [comment]
@geof90 thanks for that info. I believe i've fixed it, and have tested both code paths.

File: ReactAndroid/src/main/jni/react/jni/OnLoad.cpp [hunk 2]
```diff
@@ -696,9 +696,9 @@ static void loadScriptFromAssets(JNIEnv* env, jobject obj, jobject assetManager,
 
   env->CallStaticVoidMethod(markerClass, gLogMarkerMethod, env->NewStringUTF("loadScriptFromAssets_read"));
   if (JniJSModulesUnbundle::isUnbundle(manager, assetNameStr)) {
-    loadApplicationUnbundle(bridge, manager, script, assetNameStr);
+    loadApplicationUnbundle(bridge, manager, script, "file://" + assetNameStr);
   } else {
-    executeApplicationScript(bridge, script, assetNameStr);
+    executeApplicationScript(bridge, script, "file://" + assetNameStr);
```
Comment (mkonicek) [inline ReactAndroid/src/main/jni/react/jni/OnLoad.cpp]: This is new, correct? Why is this needed?

File: ReactAndroid/src/main/jni/react/jni/OnLoad.cpp [hunk 3]
```diff
@@ -696,9 +696,9 @@ static void loadScriptFromAssets(JNIEnv* env, jobject obj, jobject assetManager,
 
   env->CallStaticVoidMethod(markerClass, gLogMarkerMethod, env->NewStringUTF("loadScriptFromAssets_read"));
   if (JniJSModulesUnbundle::isUnbundle(manager, assetNameStr)) {
-    loadApplicationUnbundle(bridge, manager, script, assetNameStr);
+    loadApplicationUnbundle(bridge, manager, script, "file://" + assetNameStr);
```
Comment (mkonicek) [inline ReactAndroid/src/main/jni/react/jni/OnLoad.cpp]: This is new, correct? Why is this needed?

**Comment by mkonicek** [comment]
Oh those changes in OnLoad.cpp are needed for TraceKit to work even if we're loading the bundle from assets, correct?

This is a risky change but I trust you! If you're sure it won't break anything ship it :)

**Comment by mkonicek** [comment]
I should find some time to add an Android UI test for both a debug and a release build into CircleCI.

**Comment by skevy** [comment]
@mkonicek Yah so the general concept here @mkonicek is that the sourceURL argument that gets passed around really has no impact on WHAT actually gets loaded -- only how the source that is loaded is attributed.

The JS is downloaded and fetched (from either a server or from assets or from a local file) before we ever get to this point in the code. 

Without this change...the file/line number attribution in stack traces look something like this:

When loading from assets:
```
index.android.bundle:15214:16
```

When loading from a file:
```
/data/.../index.android.bundle:15214:16
```

This differs from iOS, where we get back an NSURL, no matter whether we're loading from the server or loading from the local bundle. That means that on iOS, we don't have the issue I describe, because the stack traces contain things like:

```
file://var/applications/.../index.ios.bundle:15214:16
```

Which is more inline with how browsers report stack traces. For reliable parsing of these traces, the "file://" is key. I also think that semantically it makes sense, in both loading from assets as well as loading from the local file system.

**Comment by skevy** [comment]
^^ If that makes sense to you @mkonicek, I'll ship after you give the :+1:

**Comment by mkonicek** [comment]
Cool, thanks for the explanation @skevy! So for assets, the `sourceURL` would be e.g. "file:///index.android.bundle:15214:16" and that's consistent with iOS? Sounds good, shipit :)

**Comment by skevy** [comment]
@facebook-github-bot shipit

**Comment by mkonicek** [comment]
Closed in https://github.com/facebook/react-native/commit/f2a60a202fd3dd737a968820c7941686ed1d9ddb but not automatically, weird.

**Comment by mkonicek** [comment]
Thanks for the PR @skevy!

File: ReactAndroid/src/main/jni/react/jni/OnLoad.cpp (hunk 3 above)
Comment (davidaurelio) [inline ReactAndroid/src/main/jni/react/jni/OnLoad.cpp]: Most importantly, this is wrong. Please don’t change code you can’t test.

Comment (skevy) [inline ReactAndroid/src/main/jni/react/jni/OnLoad.cpp]: I did test this. If it's broken, it's not on purpose, and I didn't change code that I couldn't test.

**Comment by davidaurelio** [comment]
The calls to `evaluateScript` here would have been the right place to add that prefix: https://github.com/skevy/react-native/blob/master/ReactAndroid/src/main/jni/react/JSCExecutor.cpp
